### PR TITLE
[DependencyInjection] Documented the ability of define the service decoration priority

### PR DESCRIPTION
**Pull request by dosten**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes (symfony/symfony#15416) |
| Applies to | 2.8+ |

### CODE REVIEW

**Comment by dupuchba** [comment]
@dosten If possible, I would write a concrete use case about this feature :-). What do you think ?

**Comment by timglabisch** [comment]
should we add a hint, that setting an absolute order, can become an antipattern?

the idea to decorate a service is to add functionality without without knowing exacly what the decorated service does. ordering means leaking such informations to the configuration.

**Comment by dosten** [comment]
@dupuchba Do you have in mind some concrete example?

@timglabisch IMO this isn't an antipattern, only define the order of the service decorators without knowing exactly what the decorated service does. i.e: The `TraceableEventDispatcher` and the `EventDispatcher`, maybe I want to add a "cache" layer to the dispatcher, but I want to apply this decorator first, so the `TraceableEventDispatcher` gets the correct memory usage, time, etc. In this example, the order don't depends of the decorated service or what it's does, but depends on another service decorator.

**Comment by dupuchba** [comment]
@dosten no ideas :( lol

**Comment by dosten** [comment]
@dupuchba IMO a concrete example here is a bad idea, an example of usage is enough.

File: components/dependency_injection/advanced.rst [hunk 1]
```diff
@@ -218,3 +218,67 @@ You can change the inner service name if you want to:
             ->addArgument(new Reference('bar.wooz'))
             ->setPublic(false)
             ->setDecoratedService('foo', 'bar.wooz');
+
+.. versionadded:: 2.8
+    The ability of define the decoration priority was introduced in 2.8.
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: [...] to define [...]

File: components/dependency_injection/advanced.rst [hunk 2]
```diff
@@ -218,3 +218,67 @@ You can change the inner service name if you want to:
...
+    Prior to Symfony 2.8, the priority depends of the order in
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: [...] depends on [...]

File: components/dependency_injection/advanced.rst [hunk 3]
```diff
@@ -218,3 +218,67 @@ You can change the inner service name if you want to:
...
+    which definitions are found.
+
+If you want to decorate the same service more than once, you can define
+the priority of decoration, the service decorator with the highest priority
+will be applied first:
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: I would reword this a bit:

> If you want to apply more than one decorator to a service, you can control their order by configuring the priority of decoration. Decorators with higher priorities will be applied first.

What do you think?

File: components/dependency_injection/advanced.rst [hunk 4]
```diff
@@ -218,3 +218,67 @@ You can change the inner service name if you want to:
...
+        $container->register('baz', 'Baz')
+            ->addArgument(new Reference('baz.inner'))
+            ->setPublic(false)
+            ->setDecoratedService('foo', null, 1);
+
+The result code will be the following:
+
+.. code-block:: php
+
+    $this->services['foo'] = new Baz(new Bar(new Foo)));
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: missing parentheses after `Foo`

**Comment by xabbuh** [comment]
Why are the decorators private while the service being decorated is public?

**Comment by dosten** [comment]
@xabbuh Comments addresed. The decorators are private to allow the `InlineServiceDefinitionsPass` to inline these services and have a real example.

**Comment by dosten** [comment]
ping @xabbuh

File: components/dependency_injection/advanced.rst [hunk 5]
```diff
@@ -219,6 +219,70 @@ You can change the inner service name if you want to:
             ->setPublic(false)
             ->setDecoratedService('foo', 'bar.wooz');
 
+.. versionadded:: 2.8
+    The ability to define the decoration priority was introduced in 2.8.
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: [...] in Symfony 2.8.

File: components/dependency_injection/advanced.rst [hunk 6]
```diff
@@ -219,6 +219,70 @@ You can change the inner service name if you want to:
...
+    Prior to Symfony 2.8, the priority depends on the order in
+    which definitions are found.
+
+If you want to apply more than one decorator to a service, you can control their
+order by configuring the priority of decoration. Decorators with higher
+priorities will be applied first.
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: you can use a colon at the end of the sentence here

File: components/dependency_injection/advanced.rst [hunk 7]
```diff
@@ -219,6 +219,70 @@ You can change the inner service name if you want to:
...
+            ->addArgument(new Reference('bar.inner'))
+            ->setPublic(false)
+            ->setDecoratedService('foo', null, 5);
+
+        $container->register('baz', 'Baz')
+            ->addArgument(new Reference('baz.inner'))
+            ->setPublic(false)
+            ->setDecoratedService('foo', null, 1);
+
+The result code will be the following:
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: I would say "The generated code will be [...]"

File: components/dependency_injection/advanced.rst [hunk 8]
```diff
@@ -219,6 +219,70 @@ You can change the inner service name if you want to:
...
+If you want to apply more than one decorator to a service, you can control their
+order by configuring the priority of decoration. Decorators with higher
+priorities will be applied first.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        foo:
+          class: Foo
```
Comment (WouterJ) [inline components/dependency_injection/advanced.rst]: please use 4 spaces for the indentation in Yaml

File: components/dependency_injection/advanced.rst [hunk 9]
```diff
@@ -219,6 +219,70 @@ You can change the inner service name if you want to:
...
+        baz:
+          class: Baz
+          public: false
+          decorates: foo
+          decoration_priority: 1
+          arguments: ["@baz.inner"]
+
+    .. code-block:: xml
+
+        <service id="foo" class="Foo" />
```
Comment (WouterJ) [inline components/dependency_injection/advanced.rst]: please use a complete XML example:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">

    <services>
        <service id="foo" class="Foo"/>

        <service id="bar" class="Bar" decorates="foo" decoration-priority="5" public="false">
            <argument type="service" id="bar.inner"/>
        </service>

        <service id="baz" class="Baz" decorates="foo" decoration-priority="1" public="false">
            <argument type="service" id="baz.inner"/>
        </service>
    </services>
</container>

**Comment by dosten** [comment]
@xabbuh @WouterJ comments addressed.

File: components/dependency_injection/advanced.rst [hunk 10]
```diff
@@ -219,6 +219,79 @@ You can change the inner service name if you want to:
...
+
+        foo:
+            class: Foo
+
+        bar:
+            class: Bar
+            public: false
+            decorates: foo
+            decoration_priority: 5
+            arguments: ["@bar.inner"]
```
Comment (xabbuh) [inline components/dependency_injection/advanced.rst]: Let's use single quotes in YAML code blocks (see #5935).

**Comment by xabbuh** [comment]
@dosten I understand why you marked the decorators as private. But what's the reason to keep the initial service public?

**Comment by dosten** [comment]
Well, if I make the initial service private the generated code is not true because if the service is private the generated code isn't `$this->services['foo'] = ...`

File: components/dependency_injection/advanced.rst [hunk 11]
```diff
@@ -219,6 +219,79 @@ You can change the inner service name if you want to:
             ->setPublic(false)
             ->setDecoratedService('foo', 'bar.wooz');
 
+.. versionadded:: 2.8
+    The ability to define the decoration priority was introduced in Symfony 2.8.
+    Prior to Symfony 2.8, the priority depends on the order in
+    which definitions are found.
+
+If you want to apply more than one decorator to a service, you can control their
+order by configuring the priority of decoration. Decorators with higher
```
Comment (javiereguiluz) [inline components/dependency_injection/advanced.rst]: I'll add a very brief mention about the allowed values for the `decoration_priority`: positive integers only? positive and negative integers?

**Comment by javiereguiluz** [comment]
@dosten thanks for this contribution. It would be nice if we could merge this PR soon because it adds docs for some important feature.

File: components/dependency_injection/advanced.rst (hunk 11 above)
Comment (dosten) [inline components/dependency_injection/advanced.rst]: Maybe:

> If you want to apply more than one decorator to a service, you can control their
order by configuring the priority of decoration, this can be any integer number (decorators with higher priorities will be applied first)

What do you think? I'm not sure if we should say "integer numbers" or "positive and negative numbers". I prefer the first one because it's short and mathematically correct.

**Comment by dosten** [comment]
@javiereguiluz @xabbuh comments addresed

**Comment by WouterJ** [comment]
:+1:

**Comment by xabbuh** [comment]
Thank you @dosten and sorry for the delay.
